### PR TITLE
/warnaserror-:DiagnosticId should not nullify editorconfig diagnostic…

### LINKED ISSUE #49446: '/warnaserror-:DiagnosticId' should not nullify diagnostic severity settings for 'DiagnosticId' in editorconfig
**Version Used**: 

**Steps to Reproduce**:

1. Create a C# class library project with `net5` target framework with below code:
```csharp
class C
{
    void M() { }
}
```
2. Verify IDE suggestion/message [CA1822: Mark members as static](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1822) is reported for method 'M'
3. Add an editorconfig file to the project root with following contents:
```editorconfig
[*.cs]
dotnet_diagnostic.CA1822.severity = warning
```
4. Build and verify `CA1822` warning
5. Edit the csproj file to enable `/warnaserror`:
```xml
<PropertyGroup>
  <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
</PropertyGroup>
```
6. Rebuild project and verify `CA1822` is reported as an error
7. Edit the csproj file to add `/warnaserror-:CA1822`:
```xml
<PropertyGroup>
  <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
  <WarningsNotAsErrors>CA1822</WarningsNotAsErrors>
</PropertyGroup>
```
8. Rebuild project

**Expected Behavior**:
`CA1822` is reported as a warning

**Actual Behavior**:
`CA1822` is not reported as a warning. `/warnaserror-:CA1822` seems to cause compiler fallback `CA1822` to its default severity `Info`/`Suggestion` and ignores editorconfig entry that promotes it to a warning.

### PR DESCRIPTION
**Pull request by mavasani**
… settings for DiagnosticId

Fixes #49446

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for second review.

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler for second review.

**Comment by AlekseyTs** [comment]
>                 if (tree != null && syntaxTreeOptions.TryGetDiagnosticValue(tree, id, cancellationToken, out report) ||

It looks like we are overriding the ```report``` value unconditionally. It feels like that can lead to a loss of the value we got from ```specificDiagnosticOptions```. Are we testing scenarios like that? I think it would be better to extract config setting into a separate variable and then have an explicit logic to merge both values.   

---
Refers to: src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs:201 in 23dbaaa. [](commit_id = 23dbaaa320b2faec8c46607b6817f336b9c33a0a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                 if (tree != null && syntaxTreeOptions.TryGetDiagnosticValue(tree, id, cancellationToken, out report) ||

For example, it is not obvious if we can switch to ```ReportDiagnostic.Error``` and whether that would be the right thing to do.

---
In reply to: [732285025](https://github.com/dotnet/roslyn/pull/49449#issuecomment-732285025) [](ancestors = 732285025)

---
Refers to: src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs:201 in 23dbaaa. [](commit_id = 23dbaaa320b2faec8c46607b6817f336b9c33a0a, deletion_comment = False)

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDiagnosticFilter.vb [hunk 1]
```diff
@@ -135,20 +135,31 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                If report = ReportDiagnostic.Default Then
+                    specifiedWarnAsErrorMinus = True
+                End If
+            End If
+
+            ' Apply syntax tree options, if applicable.
+            If syntaxTreeOptions IsNot Nothing AndAlso
+               (Not isSpecified OrElse specifiedWarnAsErrorMinus) Then
+
                 If (tree IsNot Nothing AndAlso syntaxTreeOptions.TryGetDiagnosticValue(tree, id, cancellationToken, report)) OrElse
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDiagnosticFilter.vb:155]: >report [](start = 116, length = 6)

Same comment as for C#

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 2]
```diff
@@ -12350,6 +12350,42 @@ public void TestWarnAsErrorMinusDoesNotEnableDisabledByDefaultAnalyzers(Diagnost
             }
         }
 
+        [WorkItem(49446, "https://github.com/dotnet/roslyn/issues/49446")]
+        [CombinatorialData, Theory]
+        public void TestWarnAsErrorMinusDoesNotNullifyEditorConfig(bool warnAsErrorMinus)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:12355]: >TestWarnAsErrorMinusDoesNotNullifyEditorConfig [](start = 20, length = 46)

Given the nature of the change, I think we need to test more permutations of command line settings and config settings. I would say across the existing and new tests we should cover full matrix.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

**Comment by mavasani** [comment]
> > ```
> >             if (tree != null && syntaxTreeOptions.TryGetDiagnosticValue(tree, id, cancellationToken, out report) ||
> > ```
> 
> It looks like we are overriding the `report` value unconditionally. It feels like that can lead to a loss of the value we got from `specificDiagnosticOptions`. Are we testing scenarios like that? I think it would be better to extract config setting into a separate variable and then have an explicit logic to merge both values.
> 
> Refers to: src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs:201 in [23dbaaa](https://github.com/dotnet/roslyn/commit/23dbaaa320b2faec8c46607b6817f336b9c33a0a). [](commit_id = [23dbaaa](https://github.com/dotnet/roslyn/commit/23dbaaa320b2faec8c46607b6817f336b9c33a0a), deletion_comment = False)

Thanks, I can refactor so we don't explicitly rely on the behavior of `TryGetDiagnosticValue` to set value to `ReportDiagnostic.Default` on false return case below: https://github.com/dotnet/roslyn/blob/f59f0b5ddc3db4d8e603410de79ec14b61790283/src/Compilers/Core/Portable/Compilation/SyntaxTreeOptionsProvider.cs#L77-L85
Note that this is already tested via an [existing test](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.CommandLine.UnitTests/CommandLineTests.cs,10140) which passes `/warnaserror-` without any syntax tree options, so there is no need to add a new test, but your suggested refactoring will make this more explicit.

**Comment by AlekseyTs** [comment]
> , I can refactor so we don't explicitly rely on the behavior of TryGetDiagnosticValue to set value to ReportDiagnostic.Default on false return case 

Note, this isn't the only case I am concerned about. It is not clear to me why it would be fine to read ReportDiagnostic.Error from config in our case.

---
In reply to: [732298342](https://github.com/dotnet/roslyn/pull/49449#issuecomment-732298342) [](ancestors = 732298342)

**Comment by AlekseyTs** [comment]
> Note that this is already tested via an existing test which passes /warnaserror- without any syntax tree options, so there is no need to add a new test, but your suggested refactoring will make this more explicit.

I am referring to other values that might come from config, not just lack of a value.

---
In reply to: [732304230](https://github.com/dotnet/roslyn/pull/49449#issuecomment-732304230) [](ancestors = 732304230,732298342)

**Comment by mavasani** [comment]
> > , I can refactor so we don't explicitly rely on the behavior of TryGetDiagnosticValue to set value to ReportDiagnostic.Default on false return case
> 
> Note, this isn't the only case I am concerned about. It is not clear to me why it would be fine to read ReportDiagnostic.Error from config in our case.
> 
> In reply to: [732298342](https://github.com/dotnet/roslyn/pull/49449#issuecomment-732298342) [](ancestors = 732298342)

Maybe I am not clear about the exact scenarios that you are concerned about. The logic works as follows:
1. If `SpecificDiagnosticOptions` has a value other than `ReportDiagnostic.Default`, it overrides syntax tree options from config files and determines diagnostic's effective severity. Syntax tree options from config files are ignored.
2. Otherwise, if config file options contains a value for an ID, that overrides and determines diagnostic's reported severity. This is true even if `SpecificDiagnosticOptions` has a `ReportDiagnostic.Default` from `/warnaserror-:ID`. `/warnaserror-` cannot nullify an explicit entry from config file (even `ReportDiagnostic.Error`), it can only nullify a previous `/warnaserror+` specification indicating a warning being bumped to an error.
3. If after all the above handling, diagnostic's effective severity is a warning and `/warnaserror+` is applicable for the ID, it will be bumped to an error.

I can add more tests to assert this behavior.

**Comment by AlekseyTs** [comment]
It is not obvious to me why it is desired to override ```/warnaserror-:ID``` from command line with ```/warnaserror:ID``` from config. Since you didn't test this scenario, I couldn't tell what the expectations should be

---
In reply to: [732320103](https://github.com/dotnet/roslyn/pull/49449#issuecomment-732320103) [](ancestors = 732320103)

**Comment by mavasani** [comment]
> It is not obvious to me why it is desired to override /warnaserror-:ID from command line with /warnaserror:ID from config. Since you didn't test this scenario, I couldn't tell what the expectations should be

Sounds good - I don't have a push back to change the semantics so `/warnaserror-:ID` from command line will override `ReportDiagnostic.Error` from config for warning diagnostics. IMO, these are conflicting configurations, so pretty unlikely, but your stance seems reasonable. I'll make this adjustment and add tests.

File: src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs [hunk 3]
```diff
@@ -12350,6 +12350,70 @@ public void TestWarnAsErrorMinusDoesNotEnableDisabledByDefaultAnalyzers(Diagnost
...
+        // Verify '/warnaserror-:ID' prevents escalation to 'Error' when default severity is 'Warning' and no config file setting is specified.
+        [InlineData(false, DiagnosticSeverity.Warning, null, DiagnosticSeverity.Error)]
+        [InlineData(true, DiagnosticSeverity.Warning, null, DiagnosticSeverity.Warning)]
+        // Verify '/warnaserror-:ID' prevents escalation to 'Error' when default severity is 'Warning' and config file bumps severity to 'Error'
+        [InlineData(false, DiagnosticSeverity.Warning, DiagnosticSeverity.Error, DiagnosticSeverity.Error)]
+        [InlineData(true, DiagnosticSeverity.Warning, DiagnosticSeverity.Error, DiagnosticSeverity.Warning)]
+        // Verify '/warnaserror-:ID' has no effect when default severity is 'Info' and config file bumps severity to 'Error'
+        [InlineData(false, DiagnosticSeverity.Info, DiagnosticSeverity.Error, DiagnosticSeverity.Error)]
+        [InlineData(true, DiagnosticSeverity.Info, DiagnosticSeverity.Error, DiagnosticSeverity.Error)]
+        public void TestWarnAsErrorMinusDoesNotNullifyEditorConfig(
```
Comment (mavasani) [inline src/Compilers/CSharp/Test/CommandLine/CommandLineTests.cs:12367]: @AlekseyTs Please let me know if you feel there are more combinations that would be worth testing.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 3)

**Comment by mavasani** [comment]
Thanks!

**Pull request merged**